lib/catalog/name_adjust.rb

```diff
@@ -5,16 +5,12 @@ class NameAdjust
     def self.create_copy_name(original_name, names)
       original_name.sub!(COPY_REGEX, '')
       names.select! { |name| name.match("#{COPY_REGEX} #{original_name}") }
-
-      if names.any?
-        num = get_index(names)
-        "Copy (#{num + 1}) of " + original_name
-      else
-        "Copy of " + original_name
-      end
+      copy_name(original_name, names).truncate(64)
```

Comment:
This class is used by more than one model which have different name length requirements.

Regardless of what approach we take the name length current set here (https://github.com/RedHatInsights/catalog-api/blob/master/app/models/portfolio.rb#L15) should be moved to a constant for that class.  Will need to do the same for PortfolioItem soon as well.

If we want to keep this logic completely generic (only process strings) then the parameters could take an additional (optional) parameter of length.  Maybe `max_len = nil` where `nil` defines that we do not do any truncation.

A bigger refactoring, that would make the code less generic, would be to pass the source object, instead of `original_name`, as the first parameter and assume the `original_name` comes from `object.name`.  The benefit here is that the length would not need to be passed in, but could be access from `object.class::MAX_NAME_LENGTH`.  In this case we would at least have to define `MAX_NAME_LENGTH` on both Portfolio and PortfolioItem in the same PR.

@eclarizio thoughts?

Reply:
@gmcculloug Yeah, that's a good point. Right now we don't have any sort of max length restriction on portfolio items, but it does make sense that we likely will since they display very similarly to portfolios. Anything else that might use this in the future as you pointed out could potentially have different max length requirements.

I think maybe then we change to your suggested `max_length` optional parameter, and simply not do any truncation if that doesn't exist. I agree as well that regardless of how this ends up, the `64` should likely be a constant. I think the refactoring to pass the source object feels more like it should just *be* on the source object, like maybe `portfolio.copy_name` that can then call into this with the proper parameters.

Reply:
@gmcculloug @eclarizio I've added an optional `max_length` parameter that defaults to `nil` as per your suggestion. I also decided to go the route of adding a `copy_name` method to `Portfolio` which passes the new `MAX_NAME_LENGTH` constant through to `NameAdjust.create_copy_name`. I think this approach keeps `NameAdjust` simple and maintains compatibility with existing calls.

Still need to make changes to actually _use_ this method in places where `NameAdjust` is being used however, so stay tuned.

Reply:
Hey @clacroix12, I get the idea of creating the model method but in this project we are working to keep the models thin and move the work to the service classes.  So let's drop the Portfolio method for now and just pass `MAX_NAME_LENGTH` from the `CopyPortfolio` service class.  (Note: The constant would still be defined in this model and referenced as `Portfolio::MAX_NAME_LENGTH`.)

Reply:
@gmcculloug Sorry, that was kind of my idea. I think with `NameAdjust` being in `lib`, it makes at least some sense to put a method on `Portfolio` and then call out to the "library" to do the actual calculation. It just feels a bit weird to me to have to pass in all of these parameters that an individual portfolio knows about, because that should be something a portfolio should know how to do. `NameAdjust` doesn't care what kind of object it is, just that it can...adjust names correctly.

Maybe the better solution is turning this into a mixin and then `include`ing it into the models that potentially need to have their names adjusted?

Reply:
@gmcculloug I see what you're saying now. I think I may have misunderstood before as I was thinking we were going to change the implementation of `NameAdjust` further by having it accept the portfolio object itself. I can revert my change to `CopyPortfolio` and simply pass `Portfolio::MAX_NAME_LENGTH` without the model method.